walkeeper/src/receive_wal.rs

```diff
@@ -127,19 +88,27 @@ impl<'pg> ReceiveWalConn<'pg> {
         }
 
         // if requested, ask pageserver to fetch wal from us
-        // xxx: this place seems not really fitting
-        if swh.conf.pageserver_addr.is_some() {
-            // Need to establish replication channel with page server.
-            // Add far as replication in postgres is initiated by receiver, we should use callme mechanism
-            let conf = swh.conf.clone();
-            let timelineid = swh.timeline.get().timelineid;
-            let _ = thread::Builder::new()
-                .name("request_callback thread".into())
-                .spawn(move || {
-                    request_callback(conf, timelineid, tenant_id);
-                })
-                .unwrap();
-        }
+        // as long as this wal_stream is alive, callmemaybe thread
+        // will send requests to pageserver
+        let _wal_stream = match swh.conf.pageserver_addr {
```

Comment:
If I understand it correctly here we use reference count as a criteria of liveness. It might be good to add more explanations to the comment.

Also note that the same can be achieved using Guard pattern. See ReplicationConnGuard as an example.

Am I right that we will send callmemaybe requests forever until walproposer disconnects? Is it correct?

Reply:
Thank you for the idea. Rewrote to use the guard.